src/game.cpp

```diff
@@ -8064,53 +8080,44 @@ bool game::refill_vehicle_part(vehicle &veh, vehicle_part *part, bool test)
     } else if (ftype == "plasma") {
         fuel_per_charge = 100;
     }
-    int max_fuel = part_info.size;
-    int charge_difference = (max_fuel - part->amount) / fuel_per_charge;
-    if (charge_difference < 1) {
+    int max_fuel = veh.capacity_left( part, ftype );
+    int cur_amount = veh.tank_charges( part );
+    int charge_difference = ( max_fuel - cur_amount ) / fuel_per_charge;
+    if( charge_difference < 1 ) {
         charge_difference = 1;
     }
-    bool rem_itm = min_charges <= charge_difference;
-    long used_charges = rem_itm ? min_charges : charge_difference;
-    part->amount += used_charges * fuel_per_charge;
-    if (part->amount > max_fuel) {
-        part->amount = max_fuel;
-    }
+    long used_charges = veh.tank_fill( part, ftype, p_itm->charges * fuel_per_charge );
 
     if (ftype == "battery") {
         add_msg(_("You recharge %s's battery."), veh.name.c_str());
-        if (part->amount == max_fuel) {
+        if( veh.capacity_left( part, ftype ) <= 0 ) {
             add_msg(m_good, _("The battery is fully charged."));
         }
-    } else if (ftype == "gasoline" || ftype == "diesel") {
-        add_msg(_("You refill %s's fuel tank."), veh.name.c_str());
-        if (part->amount == max_fuel) {
-            add_msg(m_good, _("The tank is full."));
-        }
-    } else if (ftype == "plutonium") {
-        add_msg(_("You refill %s's reactor."), veh.name.c_str());
-        if (part->amount == max_fuel) {
-            add_msg(m_good, _("The reactor is full."));
+    } else {
+        const auto &ftype_itype = item_controller->find_template( ftype );
+        add_msg(_("You fill %s's %s with %d units of %s."), 
+            veh.name.c_str(), veh.part_info( part ).name.c_str(), 
+            used_charges, ftype_itype->nname( 1 ).c_str() );
+        if( veh.capacity_left( part, ftype ) <= 0 ) {
+            add_msg( m_good, _("The %s is full."), veh.part_info( part ).name.c_str() );
         }
     }
 
+    used_charges = used_charges / fuel_per_charge + 
+        ( used_charges % fuel_per_charge > 0 ? 1 : 0 );
     p_itm->charges -= used_charges;
-    if (rem_itm) {
-        if (in_container) {
-            it->contents.erase(it->contents.begin());
-        } else if (&u.weapon == it) {
+    if( p_itm->charges <= 0 ) {
```

Comment:
Or simply use `player::i_rem(const item *)` and supply the item pointer `p_itm`. The function will find the item even when it's inside a container, or when it's the weapon or when it's whatever, it goes through all items in the players possessions.